src/ast/values.ts

```diff
@@ -12,3 +32,229 @@ export const UNKNOWN_EXPRESSION: ExpressionEntity = {
+		getValue: () => UNKNOWN_VALUE,
+		hasEffectsWhenAccessedAtPath: path => path.length > 1,
+		hasEffectsWhenAssignedAtPath: path => path.length > 0,
+		hasEffectsWhenCalledAtPath: path => {
+			if (path.length === 1) {
+				const subPath = path[0];
+				return isUnknownKey(subPath) || !literalMembers[type][subPath];
+			}
+			return true;
+		},
+		someReturnExpressionWhenCalledAtPath: (
+			path: ObjectPath,
+			_callOptions: CallOptions,
+			predicateFunction: SomeReturnExpressionCallback,
+			options: ExecutionPathOptions
+		) => {
+			if (path.length === 1) {
+				const subPath = path[0];
+				return isUnknownKey(subPath)
+					|| !literalMembers[type][subPath]
+					|| predicateFunction(options)(literalMembers[type][subPath].returns);
+			}
+			return true;
+		},
+		toString: () => '[[UNKNOWN STRING]]'
+	};
+}
+
+export const UNKNOWN_LITERALS = {
+	boolean: createUnknownLiteral('boolean'),
+	number: createUnknownLiteral('number'),
+	string: createUnknownLiteral('string')
+};
+const returnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: null } };
+const callsArgReturnsBoolean: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.boolean, callsArgs: [0] } };
+const returnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: null } };
+const callsArgReturnsNumber: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.number, callsArgs: [0] } };
+const returnsString: RawMemberDescription = { value: { returns: UNKNOWN_LITERALS.string, callsArgs: null } };
+
+export const UNKNOWN_OBJECT_EXPRESSION: ExpressionEntity = {
+	reassignPath: () => {},
+	forEachReturnExpressionWhenCalledAtPath: () => {},
+	getValue: () => UNKNOWN_VALUE,
+	hasEffectsWhenAccessedAtPath: path => path.length > 1,
+	hasEffectsWhenAssignedAtPath: path => path.length > 1,
+	hasEffectsWhenCalledAtPath: path => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath) || !objectMembers[subPath];
+		}
+		return true;
+	},
+	someReturnExpressionWhenCalledAtPath: (
+		path: ObjectPath,
+		_callOptions: CallOptions,
+		predicateFunction: SomeReturnExpressionCallback,
+		options: ExecutionPathOptions
+	) => {
+		if (path.length === 1) {
+			const subPath = path[0];
+			return isUnknownKey(subPath)
+				|| !objectMembers[subPath]
+				|| predicateFunction(options)(objectMembers[subPath].returns);
+		}
+		return true;
+	},
+	toString: () => '[[UNKNOWN OBJECT]]'
+};
+
+export const arrayMembers: MemberDescriptions =
+	assembleMemberDescriptions({
+		concat: returnsArray,
+		every: callsArgReturnsBoolean,
+		filter: callsArgReturnsArray,
+		find: callsArgReturnsUnknown,
+		findIndex: callsArgReturnsNumber,
+		forEach: callsArgReturnsUnknown,
+		includes: returnsBoolean,
+		indexOf: returnsNumber,
+		join: returnsString,
+		lastIndexOf: returnsNumber,
+		map: callsArgReturnsArray,
+		reduce: callsArgReturnsUnknown,
+		reduceRight: callsArgReturnsUnknown,
```

Comment:
Could `reverse` be added here as `returnsArray`?

Also would it be possible to support `shift`, `push`, `pop` and `unshift`, `sort` and `splice`?

Reply:
Unfortunately, mutating methods are much more difficult. The reason is this:
If we want to check if e.g. `.reverse()` can be called on a variable, we check if this variable has a unique value, e.g. in
```javascript
const x = [1, 2, 3]
```
`x` would have a unique value, while in
```javascript
let x = [1, 2, 3];
x = [3, 4, 5];
```
the value is not unique i.e. we are checking for reassignments. In case of a reassignment, we just internally mark the variable as "reassigned" and treat it like `undefined` since otherwise (i.e. if we were tracking all assignments), there would be a possibly exponential performance degradation (learned this the hard way).

Now if there is a unique value, we continue asking the original expression if it supports `.reverse()` until we hit the ArrayLiteral. But the algorithm no longer knows what the original variable was or any variables in between. So we do not know if any of those variables are already included in the bundle. Which is a problem because in that situation, the call would have the side-effect of mutating an included variable.

I guess its possible to find a solution for this but for now I just wanted to focus on the low-hanging fruit.

Reply:
Makes sense.